Prometheus: Don't show errors from unsuccessful API checks like rules or exemplar checks

darrenjaneczek opened the pull request:
**What this PR does / why we need it**:

The query inspector was experiencing some error pop-ups due to errors encountered on metrics datasources configured with multiple tenants.

e.g., team-a|team-b

Some of the backend implementations of these metadata endpoints aren't yet well-defined for multi-tenant queries, so they often result in error messages.

This fix suppresses those errors..

In the event of a multi-tenant datasource, these endpoints were contributing to the pop-up errors. 

`/api/v1/query_exemplars` - in a check to see if query examplars can be accessed.

`/api/v1/rules` as a metadata request
`/api/v1/metadata` as a metadata request (see: https://github.com/grafana/mimir/issues/2410)

These error messages newly became visible in version 9. Some of the underlying request errors were still present in 8.5, but they were not visible.

![image](https://user-images.githubusercontent.com/38694490/178888710-0cd49adf-777e-4481-9ba9-1c8ea40ad6f2.png)


<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->



<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

darrenjaneczek committed:
fix: multi-tenant metrics data source warnings

The query inspector was experiencing some error pop-ups due
to errors encountered on metrics datasources configured with multiple
tenants.

e.g., team-a|team-b

This fix suppresses those errors so as to not mislead the user.

darrenjaneczek commented on packages/grafana-runtime/src/utils/DataSourceWithBackend.ts:
```diff
@@ -219,14 +219,21 @@ class DataSourceWithBackend<
+  /**
+   * Get the datasource resource path for the given path
+   */
+  getResourcePath(path: string): string {
+    return `/api/datasources/${this.id}/resources/${path}`;
   }
```
The `getResourcePath` method was separated out so that it could be used by `areExamplarsAvailable`, instead of `getResource`. It uses the path to make a `fetch`, using error suppressing options.

darrenjaneczek commented on public/app/plugins/datasource/prometheus/datasource.tsx:
```diff
@@ -237,6 +237,7 @@ export class PrometheusDatasource
       this._request<T>(`/api/datasources/${this.id}/resources${url}`, params, {
         method: 'GET',
         hideFromInspector: true,
+        showErrorAlert: false,
```
This is set to showErrorAlert: false, matching the same call in the if clause above.

This suppresses the errors from the following paths:
`/api/v1/rules`
`/api/v1/metadata`

aocenas replied:
So I am just unsure here because it means that if there is any issue with labels/values API calls this will also be suppressed and user will end up with queryEditor that is failing without any indication. If we need it only for /rules and /metadata can we lift it up? Or can we just handle the error manually and look if it is a problem with multiple tenants and only hide it then?

darrenjaneczek replied:
Hmmm... true. And if we can't fetch the labels/values, it won't be immediately clear that there is a problem unless they manually write queries and execute them?

Checking for the multi-tenants case would require some backend insight. Also, is it possible to suppress the error when catching it, or is it too late?

This closed PR suggests adding a parameter (see: https://github.com/grafana/grafana/pull/52249) -- then we ensure that we only suppress on `/metadata` and `/rules`. Except, I think I'd prefer something similar to @leventebalogh's and add a parameter for `options?: Partial<BackendSrvRequest>`.

darrenjaneczek replied:
@aocenas , what is your thought on the previous "IF" block, which sets `showErrorAlert` to false already:

(See line 223 above)
```typescript
  async metadataRequest<T = any>(url: string, params = {}) {
    // If URL includes endpoint that supports POST and GET method, try to use configured method. This might fail as POST is supported only in v2.10+.
    if (GET_AND_POST_METADATA_ENDPOINTS.some((endpoint) => url.includes(endpoint))) {
      try {
        return await lastValueFrom(
          this._request<T>(`/api/datasources/${this.id}/resources${url}`, params, {
            method: this.httpMethod,
            hideFromInspector: true,
            showErrorAlert: false, // <-- this is line 223
          })
        );
      } catch (err) {
        // If status code of error is Method Not Allowed (405) and HTTP method is POST, retry with GET
        if (this.httpMethod === 'POST' && isFetchError(err) && (err.status === 405 || err.status === 400)) {
          console.warn(`Couldn't use configured POST HTTP method for this request. Trying to use GET method instead.`);
        } else {
          throw err;
        }
      }
    }
    return await lastValueFrom(
      this._request<T>(`/api/datasources/${this.id}/resources${url}`, params, {
        method: 'GET',
        hideFromInspector: true,
        showErrorAlert: false, // <-- Current line we are discussing
...        
```

aocenas replied:
> And if we can't fetch the labels/values, it won't be immediately clear that there is a problem unless they manually write queries and execute them?

Well, I think the user just won't get any autocomplete in the code editor or no options in any dropdowns in the editor. They will be just empty and if we suppress the error there is not much other indication. So it's probably rare situation but still not great not to have any message there.

So I sort of like
> This closed PR suggests adding a parameter (see: #52249) -- then we ensure that we only suppress on /metadata and /rules. Except, I think I'd prefer something similar to @leventebalogh's #51316 and add a parameter for options?: Partial<BackendSrvRequest> to metadataRequest.

for this reason, as it seems like a smaller change/less of a regression at this point.

> what is your thought on the previous "IF" block, which sets showErrorAlert to false already:

So this might need a better comment in code but the reason it is there is that at first, we do a POST request, which may fail on older Prom instances because POST is supported only from a specific version. So we try the POST and if it fails in a specific way we retry with GET. So that is why we want to handle any error from the first request manually because we can potentially recover from it, and if the second fails we just show the default message because we have nothing else to do.

darrenjaneczek commented on public/app/plugins/datasource/prometheus/datasource.tsx:
```diff
@@ -1016,11 +1017,18 @@ export class PrometheusDatasource
 
   async areExemplarsAvailable() {
     try {
-      const res = await this.getResource('/api/v1/query_exemplars', {
+      const url = this.getResourcePath('/api/v1/query_exemplars');
+      const params = {
         query: 'test',
         start: dateTime().subtract(30, 'minutes').valueOf(),
         end: dateTime().valueOf(),
-      });
+      };
+      // Avoid alerting the user if this test fails
+      const testRequest = {
+        showErrorAlert: false,
+        hideFromInspector: true,
+      };
+      const res = await lastValueFrom(getBackendSrv().fetch({ url, params, ...testRequest }));
       if (res.data.status === 'success') {
         return true;
       }
```
This reimplementation of "areExemplarsAvaialble" uses fetch to suppress the error associated with:
`/api/v1/query_exemplars`

darrenjaneczek commented on public/app/plugins/datasource/prometheus/datasource.tsx:
```diff
@@ -1016,11 +1017,20 @@ export class PrometheusDatasource
+        start: dateTime().subtract(30, 'minutes').valueOf().toString(),
+        end: dateTime().valueOf().toString(),
+      };
+      const res = await lastValueFrom(
+        this._request<PromDataSuccessResponse<PromExemplarData>>(url, params, {
+          method: 'GET',
+          // Avoid alerting the user if this test fails
+          hideFromInspector: true,
+          showErrorAlert: false,
+        })
+      );
```
This use of this._request helps us to suppress the error associated with:
/api/v1/query_exemplars

aocenas replied:
@leventebalogh how close is this PR https://github.com/grafana/grafana/pull/51316? Would probably help us here.

56quarters replied:
> This use of this._request helps us to suppress the error associated with: `/api/v1/query_exemplars`

FWIW, the `/api/v1/query_exemplars` endpoint in Mimir and GEM supports multiple tenants if enabled by configuration (since 2.0).

yesoreyeram reviewed (approved):
LGTM

bboreham commented:
Does this suppress the messages described, or all error messages?
Are errors suppressed just when multiple tenants are defined?

darrenjaneczek commented:
@bboreham 
> Does this suppress the messages described, or all error messages? Are errors suppressed just when multiple tenants are defined?

This currently suppresses all errors made through that "metadataRequest" method.

The recent PR https://github.com/grafana/grafana/pull/52249 gives finer control over this, so I will rebase and make use the parameter.

darrenjaneczek commented:
@bboreham , more specifically:

This change to `metadataRequest` will be exclusive to the `PrometheusDatasource` class.

Here are the usages:

`this.metadataRequest('/api/v1/rules');`

For `getTagKeys`
this.metadataRequest('/api/v1/labels'); 

For `getTagLabels`
this.metadataRequest(`/api/v1/label/${options.key}/values`)

Not all types of datasources (classes other than PrometheusDatasource) share the pattern of setting the following options for their `metadataRequest` methods:
```
            hideFromInspector: true,
            showErrorAlert: false,
```
Some only have `hideFromInspector: true`, and others have neither.

bboreham commented:
OK, I don't know what `metadataRequest` is.  I don't really need to know; the point is: are you confident this doesn't break anything else, or the ability to diagnose issues causing different errors?

bboreham commented:
Also I usually recommend making the PR title match what it does.
Something like "suppress all errors for metadataRequest from prometheus datasource"?

darrenjaneczek commented:
@bboreham 
> are you confident this doesn't break anything else, or the ability to diagnose issues causing different errors?

Yes, I am confident.
If there are different errors that don't just affect metadata-related calls, we will definitely see those when we start trying to make queries.

darrenjaneczek commented:
@bboreham To ensure that label value metadata fetches return an error as early as possible, I've changed the PR to only suppress errors for these three checks:

/api/v1/query_exemplars - in a check to see if query examplars can be accessed.
/api/v1/rules as a metadata request
/api/v1/metadata as a metadata request (see: https://github.com/grafana/mimir/issues/2410)

I've changed the issue title accordingly.

aocenas commented:
@darrenjaneczek merging this and adding to 9.0.5 so this also needs a backport
